third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc

```diff
@@ -768,6 +785,189 @@ function createProtocolScope(scopeId) {
   log(`Error: Initialization exception ${e}`);
 }
 
+///////////////////////////////////////////////////////////////////////////////
+// sourcemap.js
+///////////////////////////////////////////////////////////////////////////////
+
+async function registerSourceMap(ev) {
+  if (!ev.sourceMapURL) {
+    return;
+  }
+
+  const { url: sourceURL, scriptId } = ev;
+
+  let sourceBaseURL;
+  if (typeof sourceURL === "string" && isValidBaseURL(sourceURL)) {
+    sourceBaseURL = sourceURL;
+  } else if (window?.location?.href && isValidBaseURL(window?.location?.href)) {
+    sourceBaseURL = window.location.href;
+  }
+
+  let sourceMapURL;
+  try {
+    sourceMapURL = new URL(ev.sourceMapURL, sourceBaseURL).toString();
+  } catch (err) {
+    log("Failed to process sourcemap url: " + err.message);
+    return;
+  }
+
+  // If the map was a data: URL or something along those lines, we want
+  // to resolve paths in the map relative to the overall base.
+  const sourceMapBaseURL =
+    isValidBaseURL(sourceMapURL) ? sourceMapURL : sourceBaseURL;
+
+  gSourceMapData.set(scriptId, {
+    url: sourceMapURL,
+    baseUrl: sourceMapBaseURL
+  });
+
+  if (sourceMapURL.startsWith("data:")) {
+    return;
+  }
+  let sourceMap;
+  try {
+    sourceMap = await fetchText(sourceMapURL);
+  } catch (err) {
+    log(`Failed to read sourcemap ${sourceMapURL}: ${err.message}`);
+  }
+  if (!sourceMap) {
+    return;
+  }
+
+  const script = sendMessage("Debugger.getScriptSource", { scriptId });
+
+  const recordingId = getRecordingId();
+  const id = String(Math.floor(Math.random() * 10000000000));
+  const name = `sourcemap-${id}.map`;
+  const path = await writeToRecordingDirectory(name, sourceMap);
+  await addRecordingEvent(JSON.stringify({
+    kind: "sourcemapAdded",
+    path,
+    recordingId,
+    id,
+    url: sourceMapURL,
+    baseURL: sourceMapBaseURL,
+    targetContentHash: script?.scriptSource ? makeAPIHash(script.scriptSource) : undefined,
```

Comment:
FYI I realize this will also leave the hash out if the content is an empty string, which seems wrong.